### PR TITLE
feat(edge-stack): relative path support for edge stack EE-5521

### PR DESCRIPTION
**Pull request by cmenginnz**

### CODE REVIEW

File: edge/stack/stack.go [hunk 1]
```diff
@@ -115,6 +117,53 @@ func (manager *StackManager) UpdateStacksStatus(pollResponseStacks map[int]int)
...
+
+	return nil
+}
+
+func getStackFileFolder(stack *edgeStack) string {
+	stackIDStr := strconv.Itoa(int(stack.ID))
+
+	folder := path.Join(agent.EdgeStackFilesPath, stackIDStr)
+	if IsRelativePathStack(stack) {
+		folder = path.Join(stack.FilesystemPath, agent.ComposePathPrefix, stackIDStr)
```
Comment (oscarzhou-portainer) [inline edge/stack/stack.go]: use `filepath` package for both ln159 and ln161

File: edge/stack/stack.go [hunk 2]
```diff
@@ -289,6 +315,14 @@ func (manager *StackManager) performActionOnStack(queueSleepInterval time.Durati
 			return
 		}
 
+		if IsRelativePathStack(stack) {
+			dst := path.Join(stack.FilesystemPath, agent.ComposePathPrefix)
```
Comment (oscarzhou-portainer) [inline edge/stack/stack.go]: `filepath`

File: edge/stack/stack.go [hunk 3]
```diff
@@ -515,106 +559,84 @@ func (manager *StackManager) DeleteStack(ctx context.Context, stackData edge.Sta
...
+	if err != nil {
+		return err
+	}
+
+	if !deleteStack {
+		err = filesystem.PersistDir(stack.FileFolder, stackPayload.DirEntries)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (oscarzhou-portainer) [inline edge/stack/stack.go:650]: Should line 624-639 be placed above line 566?

**Review by oscarzhou-portainer — CHANGES_REQUESTED** [review]
In terms of the proposed solution, I have some concerns that I will include in the ticket.

Overall, the PR itself LGTM. There are also other few places using `path` which are not highlighted. Would recommend to use `filepath`.

File: edge/stack/stack.go (hunk 1 above)
Comment (cmenginnz) [reply edge/stack/stack.go]: fixed

File: edge/stack/stack.go (hunk 2 above)
Comment (cmenginnz) [reply edge/stack/stack.go]: fixed

File: edge/stack/stack.go (hunk 3 above)
Comment (cmenginnz) [reply edge/stack/stack.go:650]: fixed. The modification is applied to a cloned stack now.

File: exec/docker_copy.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+				fmt.Sprintf("%s:%s", composeDestination, composeDestination),
+			},
+		},
+		nil,
+		nil,
+		containerName,
+	)
+}
+
+func copyToContainer(assetPath, src, containerID, dst string) error {
```
Comment (chiptus) [inline exec/docker_copy.go]: what about using `dockerCli.CopyToContainer`? https://github.com/moby/moby/blob/master/client/container_copy.go#L33

File: docker/containers.go [hunk 5]
```diff
@@ -5,9 +5,31 @@ import (
...
+
+	err := withCli(func(cliInner *client.Client) error {
+		cli = cliInner
+		return nil
+	})
+	if err != nil {
+		return container.CreateResponse{}, err
+	}
+
+	return cli.ContainerCreate(context.Background(), config, hostConfig, networkingConfig, platform, containerName)
```
Comment (chiptus) [inline docker/containers.go]: this feels a bit unsafe and beats the purpose of `withCli` fn, isn't cli closed inside `withCli`? is there any reason why not to run these lines of code inside withCli?

File: exec/docker_copy.go (hunk 4 above)
Comment (chiptus) [reply exec/docker_copy.go]: also IMO should be inside `docker` package

File: docker/containers.go (hunk 5 above)
Comment (cmenginnz) [reply docker/containers.go]: fixed

File: exec/docker_copy.go (hunk 4 above)
Comment (cmenginnz) [reply exec/docker_copy.go]: moved to docker package. 
`dockerCli.CopyToContainer` requires tar as input. Using the docker command is simpler here.

**Review by oscarzhou-portainer — APPROVED** [review]
LGTM

**Pull request merged**